phe/util.py

```diff
@@ -160,3 +164,215 @@ def int_to_base64(source):
+    14669, 14683, 14699, 14713, 14717, 14723, 14731, 14737, 14741, 14747,
+    14753, 14759, 14767, 14771, 14779, 14783, 14797, 14813, 14821, 14827,
+    14831, 14843, 14851, 14867, 14869, 14879, 14887, 14891, 14897, 14923,
+    14929, 14939, 14947, 14951, 14957, 14969, 14983, 15013, 15017, 15031,
+    15053, 15061, 15073, 15077, 15083, 15091, 15101, 15107, 15121, 15131,
+    15137, 15139, 15149, 15161, 15173, 15187, 15193, 15199, 15217, 15227,
+    15233, 15241, 15259, 15263, 15269, 15271, 15277, 15287, 15289, 15299,
+    15307, 15313, 15319, 15329, 15331, 15349, 15359, 15361, 15373, 15377,
+    15383, 15391, 15401, 15413, 15427, 15439, 15443, 15451, 15461, 15467,
+    15473, 15493, 15497, 15511, 15527, 15541, 15551, 15559, 15569, 15581,
+    15583, 15601, 15607, 15619, 15629, 15641, 15643, 15647, 15649, 15661,
+    15667, 15671, 15679, 15683, 15727, 15731, 15733, 15737, 15739, 15749,
+    15761, 15767, 15773, 15787, 15791, 15797, 15803, 15809, 15817, 15823,
+    15859, 15877, 15881, 15887, 15889, 15901, 15907, 15913, 15919, 15923,
+    15937, 15959, 15971, 15973, 15991, 16001, 16007, 16033, 16057, 16061,
+    16063, 16067, 16069, 16073, 16087, 16091, 16097, 16103, 16111, 16127,
+    16139, 16141, 16183, 16187, 16189, 16193, 16217, 16223, 16229, 16231,
+    16249, 16253, 16267, 16273, 16301, 16319, 16333, 16339, 16349, 16361,
+    16363, 16369, 16381, 16411, 16417, 16421, 16427, 16433, 16447, 16451,
+    16453, 16477, 16481, 16487, 16493, 16519, 16529, 16547, 16553, 16561,
+    16567, 16573, 16603, 16607, 16619, 16631, 16633, 16649, 16651, 16657,
+    16661, 16673, 16691, 16693, 16699, 16703, 16729, 16741, 16747, 16759,
+    16763, 16787, 16811, 16823, 16829, 16831, 16843, 16871, 16879, 16883,
+    16889, 16901, 16903, 16921, 16927, 16931, 16937, 16943, 16963, 16979,
+    16981, 16987, 16993, 17011, 17021, 17027, 17029, 17033, 17041, 17047,
+    17053, 17077, 17093, 17099, 17107, 17117, 17123, 17137, 17159, 17167,
+    17183, 17189, 17191, 17203, 17207, 17209, 17231, 17239, 17257, 17291,
+    17293, 17299, 17317, 17321, 17327, 17333, 17341, 17351, 17359, 17377,
+    17383, 17387, 17389, 17393, 17401, 17417, 17419, 17431, 17443, 17449,
+    17467, 17471, 17477, 17483, 17489, 17491, 17497, 17509, 17519, 17539,
+    17551, 17569, 17573, 17579, 17581, 17597, 17599, 17609, 17623, 17627,
+    17657, 17659, 17669, 17681, 17683, 17707, 17713, 17729, 17737, 17747,
+    17749, 17761, 17783, 17789, 17791, 17807, 17827, 17837, 17839, 17851,
+    17863,
+]
+
+
+def miller_rabin(n, a):
+    """Run the Miller-Rabin test on n using base a"""
+    d = n-1
+    while d % 2 == 0:
+        d //= 2
+        if pow(a, d, n) == n-1:
+            return True
+    return pow(a, d, n) == 1
+
+
+def is_prime(n):
+    """Test whether n is probably prime"""
+    if n <= first_primes[-1]:
+        return n in first_primes
+    for p in first_primes:
+        if n % p == 0:
+            return False
+    for _ in range(25):
```

Comment:
Each iteration has an independent risk 4⁻¹ of not detecting a composite number; with 25 iterations, there is less than a 2⁻⁵⁰ probability that the returned number is actually composite. Since this is not something an attacker can target in common cases (i.e. initial key generation), that should be plenty enough. However, if we want to be paranoid in case this is used in other context, we can up it to 40 to get the standard 2⁻⁸⁰ bound on failure probability.

I have added comments summing this up.